From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Benjamin Marzinski <bmarzins@redhat.com>
Date: Thu, 18 Sep 2025 19:29:08 -0400
Subject: [PATCH] libmpathpersist: Handle RESERVE with reservation held by
 failed path

Issuing a RESERVE on a device that already holds the reservation should
succeed, as long as the type is the same. But if the path that holds the
reservation is unavailable, mpathpersist fails, since it gets a
reservation conflict on all available paths. To deal with this, if the
multipath device has failed paths, and the key holding the reservation
matches the multipath device's key, and multipathd says that it is
holding the reservation, assume the reservation is held by a failed path
and claim the RESERVE succeeded, even though none of the actual scsi
commands did

Signed-off-by: Benjamin Marzinski <bmarzins@redhat.com>
---
 libmpathpersist/mpath_persist.c | 48 +++++++++++++++++++++++++--------
 1 file changed, 37 insertions(+), 11 deletions(-)

diff --git a/libmpathpersist/mpath_persist.c b/libmpathpersist/mpath_persist.c
index 6a254d2e..fce28723 100644
--- a/libmpathpersist/mpath_persist.c
+++ b/libmpathpersist/mpath_persist.c
@@ -104,7 +104,7 @@ int libmpathpersist_exit(void)
 static int mpath_prout_common(struct multipath *mpp,int rq_servact,
 			      int rq_scope, unsigned int rq_type,
 			      struct prout_param_descriptor* paramp, int noisy,
-			      struct path **pptr);
+			      struct path **pptr, bool *failed_paths);
 
 int
 mpath_prin_activepath (struct multipath *mpp, int rq_servact,
@@ -286,12 +286,12 @@ int __mpath_persistent_reserve_in (int fd, int rq_servact,
 }
 
 static int reservation_key_matches(struct multipath *mpp, uint8_t *key,
-				   int noisy)
+				   unsigned int *type)
 {
 	struct prin_resp resp = {0};
 	int status;
 
-	status = mpath_prin_activepath (mpp, MPATH_PRIN_RRES_SA, &resp, noisy);
+	status = mpath_prin_activepath(mpp, MPATH_PRIN_RRES_SA, &resp, 0);
 	if (status != MPATH_PR_SUCCESS) {
 		condlog(0, "%s: pr in read reservation command failed.",
 			mpp->wwid);
@@ -299,8 +299,12 @@ static int reservation_key_matches(struct multipath *mpp, uint8_t *key,
 	}
 	if (!resp.prin_descriptor.prin_readresv.additional_length)
 		return YNU_NO;
-	if (memcmp(key, resp.prin_descriptor.prin_readresv.key, 8) == 0)
+	if (memcmp(key, resp.prin_descriptor.prin_readresv.key, 8) == 0) {
+		if (type)
+			*type = resp.prin_descriptor.prin_readresv.scope_type &
+				MPATH_PR_TYPE_MASK;
 		return YNU_YES;
+	}
 	return YNU_NO;
 }
 
@@ -317,11 +321,20 @@ static void set_ignored_key(struct multipath *mpp, uint8_t *curr_key,
 		return;
 	if (get_prhold(mpp->alias) == PR_UNSET)
 		return;
-	if (reservation_key_matches(mpp, curr_key, 0) == YNU_NO)
+	if (reservation_key_matches(mpp, curr_key, NULL) == YNU_NO)
 		return;
 	memcpy(key, curr_key, 8);
 }
 
+static bool check_holding_reservation(struct multipath *mpp, unsigned int *type)
+{
+	if (get_be64(mpp->reservation_key) &&
+	    get_prhold(mpp->alias) == PR_SET &&
+	    reservation_key_matches(mpp, (uint8_t *)&mpp->reservation_key, type) == YNU_YES)
+		return true;
+	return false;
+}
+
 static int preempt_self(struct multipath *mpp, int rq_servact, int rq_scope,
 			unsigned int rq_type, int noisy, bool do_release)
 {
@@ -349,6 +362,7 @@ static int do_mpath_persistent_reserve_out(vector curmp, vector pathvec, int fd,
 	struct config *conf;
 	bool unregistering, preempting_reservation = false;
 	bool updated_prkey = false;
+	bool failed_paths = false;
 
 	ret = mpath_get_map(curmp, fd, &alias, &mpp);
 	if (ret != MPATH_PR_SUCCESS)
@@ -436,7 +450,7 @@ static int do_mpath_persistent_reserve_out(vector curmp, vector pathvec, int fd,
 		break;
 	case MPATH_PROUT_PREE_SA :
 	case MPATH_PROUT_PREE_AB_SA :
-		if (reservation_key_matches(mpp, paramp->sa_key, noisy) == YNU_YES) {
+		if (reservation_key_matches(mpp, paramp->sa_key, NULL) == YNU_YES) {
 			preempting_reservation = true;
 			if (memcmp(paramp->sa_key, &zerokey, 8) == 0) {
 				/* all registrants case */
@@ -453,10 +467,18 @@ static int do_mpath_persistent_reserve_out(vector curmp, vector pathvec, int fd,
 		}
 		/* fallthrough */
 	case MPATH_PROUT_RES_SA :
-	case MPATH_PROUT_CLEAR_SA:
+	case MPATH_PROUT_CLEAR_SA: {
+		unsigned int res_type;
 		ret = mpath_prout_common(mpp, rq_servact, rq_scope, rq_type,
-					 paramp, noisy, NULL);
+					 paramp, noisy, NULL, &failed_paths);
+		if (rq_servact == MPATH_PROUT_RES_SA &&
+		    ret != MPATH_PR_SUCCESS && failed_paths &&
+		    check_holding_reservation(mpp, &res_type) &&
+		    res_type == rq_type)
+			/* The reserve failed, but multipathd says we hold it */
+			ret = MPATH_PR_SUCCESS;
 		break;
+	}
 	case MPATH_PROUT_REL_SA:
 		ret = mpath_prout_rel(mpp, rq_servact, rq_scope, rq_type, paramp, noisy);
 		break;
@@ -614,7 +636,7 @@ void preempt_missing_path(struct multipath *mpp, uint8_t *key, uint8_t *sa_key,
 	memcpy(paramp.key, sa_key, 8);
 	memcpy(paramp.sa_key, key, 8);
 	status = mpath_prout_common(mpp, MPATH_PROUT_PREE_SA, rq_scope, rq_type,
-				    &paramp, noisy, NULL);
+				    &paramp, noisy, NULL, NULL);
 	if (status != MPATH_PR_SUCCESS)
 		condlog(0, "%s: register: pr preempt command failed.",
 			mpp->wwid);
@@ -853,7 +875,7 @@ void * mpath_prout_pthread_fn(void *p)
 static int mpath_prout_common(struct multipath *mpp,int rq_servact, int rq_scope,
 			      unsigned int rq_type,
 			      struct prout_param_descriptor* paramp, int noisy,
-			      struct path **pptr)
+			      struct path **pptr, bool *failed_paths)
 {
 	int i,j, ret;
 	struct pathgroup *pgp = NULL;
@@ -866,6 +888,8 @@ static int mpath_prout_common(struct multipath *mpp,int rq_servact, int rq_scope
 			if (!((pp->state == PATH_UP) || (pp->state == PATH_GHOST))){
 				condlog (1, "%s: %s path not up. Skip",
 					 mpp->wwid, pp->dev);
+				if (failed_paths)
+					*failed_paths = true;
 				continue;
 			}
 
@@ -899,6 +923,8 @@ static int mpath_prout_common(struct multipath *mpp,int rq_servact, int rq_scope
 			}
 			if (ret != MPATH_PR_RETRYABLE_ERROR)
 				return ret;
+			if (failed_paths)
+				*failed_paths = true;
 		}
 	}
 	if (found)
@@ -933,7 +959,7 @@ static int do_preempt_self(struct multipath *mpp, struct be64 sa_key,
 	memcpy(paramp.key, &mpp->reservation_key, 8);
 	memcpy(paramp.sa_key, &sa_key, 8);
 	status = mpath_prout_common(mpp, rq_servact, rq_scope, rq_type,
-				    &paramp, noisy, &pp);
+				    &paramp, noisy, &pp, NULL);
 	if (status != MPATH_PR_SUCCESS) {
 		condlog(0, "%s: self preempt command failed.", mpp->wwid);
 		goto fail_resume;
